### PR TITLE
frontend: add relationship between actions and projects/builds

### LINKED ISSUE #2823: add relationship between actions and projects/builds
tracker issue of #2805

### PR DESCRIPTION
**Pull request by nikromen**
This will help to implement RFEs like #1108

TODO:
- [x] discuss what to do with `object_type` (e.g. what is None?)
- [x] do migration when we agree on relationships between tables
- [x] ~~tests~~ This PR just connects Copr and Action model, write tests for the RFEs like #1108

~~Merge after #2733 and change `down_revision` to `7d9f6f921fa0`~~ done

Closes #2823

### CODE REVIEW

**Comment by FrostyX** [comment]
I am setting blocked so we don't accidentally merge before this

> Merge after https://github.com/fedora-copr/copr/pull/2733 and change down_revision to 7d9f6f921fa0

**Comment by praiskup** [comment]
Note that #2826 adds yet another migration script.

File: frontend/coprs_frontend/coprs/models.py [hunk 1]
```diff
@@ -2139,6 +2139,9 @@ class Action(db.Model, helpers.Serializer):
     # time ended as returned by int(time.time())
     ended_on = db.Column(db.Integer, index=True)
 
+    copr_id = db.Column(db.Integer, db.ForeignKey("copr.id"), nullable=True)
```
Comment (praiskup) [inline frontend/coprs_frontend/coprs/models.py:2142]: I think we should index this field.

File: frontend/coprs_frontend/coprs/models.py [hunk 2]
```diff
@@ -2139,6 +2139,9 @@ class Action(db.Model, helpers.Serializer):
...
+    copr = db.relationship("Copr", backref=db.backref("actions"), foreign_keys=[copr_id])
```
Comment (praiskup) [inline frontend/coprs_frontend/coprs/models.py:2143]: The `foreign_keys` is imo redundant.

File: frontend/coprs_frontend/coprs/logic/coprs_logic.py [hunk 3]
```diff
@@ -857,13 +857,14 @@ def send_delete_dirs_action(cls):
...
-
-        db.session.add(action)
+            action = models.Action(
+                action_type=ActionTypeEnum("remove_dirs"),
+                object_type="copr",
+                data=json.dumps(remove_dirs),
+                created_on=int(time.time()),
+                copr_id=copr_id,
+            )
+            db.session.add(action)
```
Comment (praiskup) [inline frontend/coprs_frontend/coprs/logic/coprs_logic.py:867]: This is weird, is the indent change intentional?

File: frontend/coprs_frontend/alembic/versions/77a33109d365_connect_actions_with_copr_project.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+# revision identifiers, used by Alembic.
+revision = '77a33109d365'
+down_revision = 'ba6ac0936bfb'
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    op.add_column('action', sa.Column('copr_id', sa.Integer(), nullable=True))
+    op.create_foreign_key(None, 'action', 'copr', ['copr_id'], ['id'])
```
Comment (praiskup) [inline frontend/coprs_frontend/alembic/versions/77a33109d365_connect_actions_with_copr_project.py:22]: note for myself: check this is synced

File: frontend/coprs_frontend/coprs/logic/actions_logic.py [hunk 5]
```diff
@@ -246,7 +249,9 @@ def send_delete_multiple_builds(cls, builds):
             action_type=ActionTypeEnum("delete"),
             object_type="builds",
             data=json.dumps(data),
-            created_on=int(time.time())
+            created_on=int(time.time()),
+            # TODO: all builds are from the same Copr?
```
Comment (praiskup) [inline frontend/coprs_frontend/coprs/logic/actions_logic.py:253]: Yes.

**Comment by praiskup** [comment]
> I am setting blocked so we don't accidentally merge before this
> 
> > Merge after #2733 and change down_revision to 7d9f6f921fa0

I actually think that we need to merge in vice-versa order :-) I just created a better tag for this than `blocked`.

**Comment by praiskup** [comment]
Ah, no - #2733 is different.

File: frontend/coprs_frontend/coprs/logic/coprs_logic.py (hunk 3 above)
Comment (praiskup) [reply frontend/coprs_frontend/coprs/logic/coprs_logic.py:867]: Hmm, this sounds good, thanks!

File: frontend/coprs_frontend/alembic/versions/08dd42f4c304_connect_copr_to_actions.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+Create Date: 2023-08-03 15:45:53.527538
+"""
+
+from alembic import op
+import sqlalchemy as sa
+
+
+# revision identifiers, used by Alembic.
+revision = '08dd42f4c304'
+down_revision = 'daa62cd0743d'
```
Comment (praiskup) [inline frontend/coprs_frontend/alembic/versions/08dd42f4c304_connect_copr_to_actions.py:14]: I updated this.

**Review by praiskup — APPROVED** [review]
Nice work, thanks!

**Pull request merged**